libs/langchain-google-common/src/utils/media_core.ts

```diff
@@ -0,0 +1,465 @@
+      delete newBlob.path;
+    }
+    return newBlob;
+  }
+
+  async _validBlobPrefixPath(blob: MediaBlob, opts?: BlobStoreStoreOptions): Promise<MediaBlob> {
+    const prefix = opts?.pathPrefix ?? "";
+    const suffix = this._blobPathSuffix(blob);
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  async _validBlobPrefixUuid(blob: MediaBlob, opts?: BlobStoreStoreOptions): Promise<MediaBlob> {
+    const prefix = opts?.pathPrefix ?? "";
+    const suffix = uuidv4();   // TODO - option to specify version?
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  async _validBlobRemovePath(blob: MediaBlob, _opts?: BlobStoreStoreOptions): Promise<MediaBlob> {
+    return this._newBlob(blob, undefined);
+  }
+
+  /**
+   * Based on the blob and options, return a blob that has a valid path
+   * that can be saved.
+   * @param blob
+   * @param opts
+   */
+  async _validStoreBlob(blob: MediaBlob, opts?: BlobStoreStoreOptions): Promise<MediaBlob | undefined> {
+    if (await this._hasValidPath(blob, opts)) {
+      return blob;
+    }
+    switch (opts?.actionIfInvalid) {
+      case "ignore": return blob;
+      case "prefixPath": return this._validBlobPrefixPath(blob, opts);
+      case "prefixUuid": return this._validBlobPrefixUuid(blob, opts);
+      case "removePath": return this._validBlobRemovePath(blob, opts);
+      default: return undefined;
+    }
+  }
+
+  async store(blob: MediaBlob, opts: BlobStoreStoreOptions = {}): Promise<MediaBlob | undefined> {
+    const allOpts: BlobStoreStoreOptions = {...this.defaultStoreOptions, ...opts};
+    const validBlob = await this._validStoreBlob(blob, allOpts);
+    if (typeof validBlob !== "undefined") {
+      const validKey = await validBlob.asUri();
+      await this.mset([[validKey, validBlob]]);
+      const savedKey = await validBlob.asUri();
+      return (await this.fetch(savedKey));
+    }
+    return undefined;
+  }
+
+  async _missingFetchBlobEmpty(path: string, _opts?: BlobStoreFetchOptions): Promise<MediaBlob> {
+    return new MediaBlob({path});
+  }
+
+  async _missingFetchBlob(path: string, opts?: BlobStoreFetchOptions): Promise<MediaBlob | undefined> {
+    switch (opts?.actionIfBlobMissing) {
+      case "emptyBlob": return this._missingFetchBlobEmpty(path, opts);
+      default: return undefined;
+    }
+  }
+
+  async fetch(key: string | MediaBlob, opts: BlobStoreFetchOptions = {}): Promise<MediaBlob | undefined> {
+    const allOpts: BlobStoreFetchOptions = {...this.defaultFetchOptions, ...opts};
+    const realKey = await this._realKey(key);
+    const ret = await this.mget([realKey]);
+    return ret?.[0] ?? await this._missingFetchBlob(realKey, allOpts);
+  }
+}
+
+export interface BackedBlobStoreOptions extends BlobStoreOptions {
+  backingStore: BaseStore<string, MediaBlob>;
+}
+
+export class BackedBlobStore extends BlobStore {
+  backingStore: BaseStore<string, MediaBlob>;
```

Comment:
Gut feeling is this should be part of the base class

Reply:
The `backingStore` attribute?
Not all of the implementations need a separate way to store the data, since some can do it on their own. (For example, a BlobStore for Google Cloud Storage doesn't - it can store both data and metadata for a URL for the store.)
The `BackedBlobStore` class is mostly a way to have a BlobStore that enforces the BlobStore semantics of store/fetch but that is backed by other generic store systems that can store a serialized `MediaBlob` in... whatever permanent means is appropriate without having to write custom versions of all of them.

Reply:
(I guess I wouldn't object if it's part of the `BlobStore`. But there are classes where setting it makes no sense.)